source/includes/inkling2-reference/_types.html.md

```diff
@@ -0,0 +1,159 @@
+  Number.Bool
+```
+
+Inkling supports built-in symbolic names for number types that correspond to integer and floating-point types commonly found in other typed languages. These are shortands for constrained number types.
+
+These types make use of the built-in namespace `Number`, which must be referenced somewhere in the program with a `using` statement.
+
+*	Unsigned integers: UInt8, UInt16, UInt32, UInt64
+*	Signed Integers: Int8, Int16, Int32, Int64
+*	Floating-point: Float32, Float64
+*	Boolean: Bool
+
+### String Constraints
+
+```inkling2--code
+  string<"A", "B", "C">
+  string<"Z">
+  "Z"
+  number<Left = "L", Straight = "S", Right = "R">
+```
+
+As with numbers, a `string` type can be constrained using an enumerated constraint.
+
+### Complex Types
+
+```inkling2--code
+  using Image
+```
+
+```inkling2--code
+  Image.Gray<100, 200>
+  Image.Gray<Height=200, Width=100>
+```
+
+Currently, the only supported complex type is Image.Gray, which represents a two-dimensional collection of gray-scale pixels. Each pixel is represented as a Number.UInt8 value (0 is black and 255 is white).
+
+The width and height of the image must be specified using a constant expression.
+
+## Type Declarations
+```inkling2--code
+  type Height number
+
+  type Coordinate {X: number, Y: number}
+
+  type TicTacToeValue string<"", "X", "O">
+
+  type TicTacToeBoard TicTacToeValue[3][3]
+```
+
+A type can be associated with an identifier using a type declaration statement. This identifier can be used anywhere a type literal would be used.
+
+> Type Declaration Syntax
+
+```inkling2--syntax
+  typeDeclaration :==
+    type <typeName> typeReference
+
+  typeReference :==
+    <typeName> | typeLiteral
+
+  typeLiteral :==
+    primitiveType |
+    typeReference'<' rangeExpression '>' |
+    typeReference'<' valueList '>' |
+    typeReference'<' typeArgumentList '>' |
+    typeReference '[' constExpression ']' |
+    '{' typeFieldList '}'
+
+  rangeExpression :==
+    constExpression '..' constExpression [step constExpression]?
+
+  valueList :==
+    constExpression [',' constExpression]* [',']?
+
+  typeArgumentList :==
+    typeArgument [',' typeArgument]* [',']?
+
+  typeArgument :==
+    [<paramName> '=']? typeReference
+
+  typeFieldList :==
+    typeField [',' typeField]* [',']?
+
+  typeField :==
+    <fieldName> ':' typeReference
+```
+
+## Type Equivalence
+
+Inkling uses types to understand and interpret the format of data. The Inkling compiler performs type checks to guarantee type compatibility. In some cases, type equivalence is required (i.e. the types must match exactly). In other cases, castability is required (i.e. one type can be trivially transformed into another type).
+
+The table below explains the rules for equivalence and castability.
+
+  Type      |  Rules for equivalence  |  Rules for castability
+----------- | ----------------------- | ----------------------- 
+Number | Constraints match | Source is more constrained than dest
```

Comment:
Could do with an example, 'more constrained' has different meanings for nominal, ordinal.